design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,190 @@
+```
+
+The `SHOW TYPES` command will list all custom types and their type definitions. A sample output
+for this command would be:
+```
+|---------------|--------------------------------------------------------------|------------------|
+|     name      |       definition                                             | source           |
+|---------------|--------------------------------------------------------------|------------------|
+| ADDRESS       | STRUCT<number INTEGER, street VARCHAR, city VARCHAR>         | types.sql        |
+| PERSON        | STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS> | types.sql        |
+| COMPANY       | STRUCT<name VARCHAR, headquarters ADDRESS>                   | CLI              |
+|---------------|--------------------------------------------------------------|------------------|
+```
+
+### Extension Directory
+
+The `ksql.extension.dir` will now also recognize `.sql` files that contain only `CREATE TYPE`
+commands. These commands will all be run before compiling any UDFs so these custom type declarations
+can be used in the `@Udf` annotations. They will be loaded in natural order to ensure that the
+loading behavior is deterministic.
+
+## Design
+
+There will be a rewriter phase that will replace any aliased type with the full schema. For UDFs, 
+the annotation will lookup the schema directly and compile the UDF using the full schema. There
+are no restrictions on what types can be added as custom types, and custom types can be composite.
+
+This rewriting will be done _before_ enqueuing the command on the command topic to make sure that
+if the types change the existing statements will already be resolved. The user will see a success
+message that contains the rewritten schema.
+
+If a type is already registered via CLI, but is then added to the extensions directory, the value
+in the directory will take precedence over the CLI registered value. The CLI will reject calls to
+register types that are already 
+
+## Future Work
+* in v1, `DESCRIBE`/`Explian` commands would show the flattened types (resolved) - it is better to 
+keep the original type structure and map it back for `DESCRIBE`/`Explain` commands
+* SchemaRegistry integration is not in scope
+
+## Test plan
+
+Nothing special of note here.
+
+## Documentation Updates
+
+* The `syntax-reference.rst` will have an updated section on `CREATE TYPE`:
+
+>**Synopsis**
+>
+>.. code:: sql
+>
+>    CREATE TYPE <type_name> AS <type>;
+>
+> Example:
+>
+>.. code:: sql
+>
+>   CREATE TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;
+>   CREATE TYPE PERSON AS STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS>;
+>   CREATE TYPE COMPANY AS STRUCT<name VARCHAR, headquarters ADDRESS>;
+>
+>**Description**
+>
+>Register the ``<type>`` under the alias ``<type_name>`` to be used in future statements. Any valid
+>column schema can be registered as a custom type, and the type_name must be exclusively letters
+>```[A-Z]```.
```

Comment:
Is there a reason we're not allowing a delimiter (like the underscore) in type names? If we're limited to uppercase letters only, multi-word names become tough to read.

Reply:
I just wanted it to be simple (perhaps avoid parsing errors). I will look to see if this is complicated to support